### PR TITLE
sign.c, PVS static analysis/ removed dead code branches

### PR DESCRIPTION
**Pull request by bartop**

### CODE REVIEW

**Comment by blueyed** [comment]
Thanks!
Just for reference: at least the first part came into nvim via a Vim patch (but was adjusted) in 83025f002.

**Comment by blueyed** [comment]
@bartop 
Have you compared the new code to what Vim currently has there?

**Comment by bartop** [comment]
No I haven't, I will look into it in a moment

**Comment by bartop** [comment]
@blueyed I checked it, they use in vim different allocation function now. It's called 'lalloc_id' or something like this: https://github.com/vim/vim/blob/4e038571aa91521e110187a256b5d16bff8b5820/src/sign.c#L204

**Comment by bartop** [comment]
Btw, did I break something on travis or is it some randomized error from ASAN?

**Comment by blueyed** [comment]
> Btw, did I break something on travis or is it some randomized error from ASAN?

Restarted - something flaky (https://github.com/neovim/neovim/issues/8813).

**Comment by blueyed** [comment]
Yeah, there are two Vim patches touching this code:

- [patch 8.1.1386: unessesary type casts for lalloc()](https://github.com/vim/vim/commit/18a4ba29aeccb9841d5bfdd2eaaffdfae2f15ced)
- [patch 8.1.1414: alloc() returning "char_u *" causes a lot of type casts](https://github.com/vim/vim/commit/c799fe206e61f2e2c1231bc46cbe4bb354f3da69)

While still having not looked closer at it, it might make sense to port those first / include them here.

**Comment by bartop** [comment]
So I took a look a those two patches you mentioned and it seems that they either are not applicable or not applicable yet. There is currently no typecasts in the places the patches remove them from.

File: src/nvim/sign.c [hunk 1]
```diff
@@ -1374,16 +1365,15 @@ void sign_getlist(const char_u *name, list_T *retlist)
 list_T *get_buffer_signs(buf_T *buf)
   FUNC_ATTR_NONNULL_RET FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
 {
-  signlist_T  *sign;
-  dict_T  *d;
-  list_T *const l = tv_list_alloc(kListLenMayKnow);
+  signlist_T *sign;
+  dict_T *dict;
+  list_T *const list = tv_list_alloc(kListLenMayKnow);
```
Comment (justinmk) [inline src/nvim/sign.c]: There are at least 2 reason not to rename things like this:

1. The scope is very small. Using long names for small scopes is more harmful than useful.
2. It is unidiomatic for the Vim/Nvim codebase. We are not going to rename every `d` to `dict`, and it doesn't add useful information anyway. `d` is almost always a dict. `l` is very often a list.

**Comment by justinmk** [comment]
LGTM after reverting the variable renaming.

File: src/nvim/sign.c (hunk 1 above)
Comment (bartop) [reply src/nvim/sign.c]: Ok, I am not fully convinced (entering codebase with variable `d` vs `dict` makes difference for me), but I reverted the naming changes ;)

Comment (justinmk) [reply src/nvim/sign.c]: the rename is still present

Comment (bartop) [reply src/nvim/sign.c]: @justinmk Now should be good

File: src/nvim/sign.c [hunk 2]
```diff
@@ -1419,9 +1404,7 @@ static void sign_get_placed_in_buf(
         || (sign_id == 0 && lnum == sign->lnum)
         || (lnum == 0 && sign_id == sign->id)
         || (lnum == sign->lnum && sign_id == sign->id)) {
-      if ((sdict = sign_get_info(sign)) != NULL) {
-        tv_list_append_dict(l, sdict);
-      }
+      tv_list_append_dict(l, sign_get_info(sign));
```
Comment (justinmk) [inline src/nvim/sign.c]: using local vars like `sdict` is good practice for e.g. debugging. Next time I would just leave it.

**Comment by justinmk** [comment]
Thanks. See also https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pvs-studio , and check `git log --grep PVS` for previous examples.

**Pull request merged**